internal/appsec/emitter/waf/metrics.go

```diff
@@ -0,0 +1,288 @@
+func (m *ContextMetrics) RegisterWafRun(addrs waf.RunAddressData, tags RequestMilestones) {
+	switch addrs.Scope {
+	case waf.RASPScope:
+		m.SumRASPCalls.Add(1)
+		ruleType, ok := addresses.RASPRuleTypeFromAddressSet(addrs)
+		if !ok {
+			telemetrylog.Error("unexpected call to RASPRuleTypeFromAddressSet", telemetry.WithTags([]string{"product:appsec"}))
+			return
+		}
+		if metric, ok := m.raspRuleEval[ruleType]; ok {
+			metric.Submit(1)
+		}
+		if tags.ruleTriggered {
+			blockTag := "block:irrelevant"
+			if tags.requestBlocked { // TODO: add block:failure to the mix
+				blockTag = "block:success"
+			}
+			telemetry.Count(telemetry.NamespaceAppSec, "rasp.rule.match", append([]string{
+				blockTag,
+			}, m.baseRASPTags[ruleType]...)).Submit(1)
+		}
+	case waf.DefaultScope, "":
+		if tags.requestBlocked {
+			m.Milestones.requestBlocked = true
+		}
+		if tags.ruleTriggered {
+			m.Milestones.ruleTriggered = true
+		}
+		if tags.wafTimeout {
+			m.Milestones.wafTimeout = true
+		}
+		if tags.rateLimited {
+			m.Milestones.rateLimited = true
+		}
+		if tags.wafError {
+			m.Milestones.wafError = true
+		}
+	default:
+		telemetrylog.Error("unexpected scope name: %v", addrs.Scope, telemetry.WithTags([]string{"product:appsec"}))
+	}
+}
+
+// IncWafError should be called if go-libddwaf.(*Context).Run() returns an error to increments metrics linked to WAF errors
+func (m *ContextMetrics) IncWafError(addrs waf.RunAddressData, in error) {
+	if in == nil {
+		return
+	}
+
+	if !errors.Is(in, wafErrors.ErrTimeout) {
+		telemetrylog.Error("unexpected WAF error: %v", in, telemetry.WithTags(append([]string{
+			"product:appsec",
+		}, m.baseTags...)))
+	}
+
+	switch addrs.Scope {
+	case waf.RASPScope:
+		ruleType, ok := addresses.RASPRuleTypeFromAddressSet(addrs)
+		if !ok {
+			telemetrylog.Error("unexpected call to RASPRuleTypeFromAddressSet: %v", in, telemetry.WithTags([]string{"product:appsec"}))
+		}
+		m.raspError(in, ruleType)
+	case waf.DefaultScope, "":
+		m.wafError(in)
+	default:
+		telemetrylog.Error("unexpected scope name: %v", addrs.Scope, telemetry.WithTags([]string{"product:appsec"}))
+	}
+}
+
+func (m *ContextMetrics) wafError(in error) {
+	m.SumWAFErrors.Add(1)
+	errCode := -127 // Default error code if the error does not implement [waf.RunError]
+	if code := wafErrors.ToWafErrorCode(in); code != 0 {
+		errCode = code
+	}
+
+	telemetry.Count(telemetry.NamespaceAppSec, "waf.error", append([]string{
+		"error_code:" + strconv.Itoa(errCode),
+	}, m.baseTags...)).Submit(1)
+}
+
+func (m *ContextMetrics) raspError(in error, ruleType addresses.RASPRuleType) {
+	m.SumRASPErrors.Add(1)
+	errCode := -127 // Default error code if the error does not implement [waf.RunError]
+	if code := wafErrors.ToWafErrorCode(in); code != 0 {
+		errCode = code
+	}
+
+	telemetry.Count(telemetry.NamespaceAppSec, "rasp.error", append([]string{
+		"error_code:" + strconv.Itoa(errCode),
+	}, m.baseRASPTags[ruleType]...)).Submit(1)
+}
```

Comment:
Just moving the magic number into a constant...

```suggestion
const defaultErrorCode = -127 // Default error code if the error does not implement [waf.RunError]

func (m *ContextMetrics) wafError(in error) {
	m.SumWAFErrors.Add(1)
	errCode := defaultErrorCode 
	if code := wafErrors.ToWafErrorCode(in); code != 0 {
		errCode = code
	}

	telemetry.Count(telemetry.NamespaceAppSec, "waf.error", append([]string{
		"error_code:" + strconv.Itoa(errCode),
	}, m.baseTags...)).Submit(1)
}

func (m *ContextMetrics) raspError(in error, ruleType addresses.RASPRuleType) {
	m.SumRASPErrors.Add(1)
	errCode := defaultErrorCode
	if code := wafErrors.ToWafErrorCode(in); code != 0 {
		errCode = code
	}

	telemetry.Count(telemetry.NamespaceAppSec, "rasp.error", append([]string{
		"error_code:" + strconv.Itoa(errCode),
	}, m.baseRASPTags[ruleType]...)).Submit(1)
}
```